### PR TITLE
Instructions on how to run wire-server with fake AWS services

### PR DESCRIPTION
**Pull request by tiago-loureiro**
The intent of this PR is to allow anyone to set up wire-server using `docker` to set up dependencies while running the different services in `wire-server` in a native fashion. It includes:

 * instructions :)
 * bash scripts that automate starting up services
 * automatic generation of secrets
 * some sensible default configuration for all services
 * swagger setup

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -19,12 +19,13 @@ fast: init
...
 	stack clean
+	$(MAKE) -C services/nginz clean
 	-rm -rf dist
 	-rm -f .metadata
 
 .PHONY: services
-services:
-	$(foreach service,$(HASKELL_SERVICES),$(MAKE) -C services/$(service);)
+services: init
+	$(foreach service,$(SERVICES),$(MAKE) -C services/$(service);)
```
Comment (jschaul) [inline Makefile]:
```Makefile
services: init install
	$(MAKE) -C services/nginz
```
?

File: services/nginz/conf/common_response.conf [hunk 2]
```diff
@@ -0,0 +1,31 @@
+      # Should be overriden when using websockets
```
Comment (jschaul) [inline services/nginz/conf/common_response.conf]: Let's move these configuration values (or includes thereof) to `deploy/services-demo` as they're unrelated to nginz as such, they only make sense in the context of all of wire-server.

File: deploy/services-demo/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# How to run `wire-server` with "fake" AWS dependencies
+
+This document assumes that you have already compiled all services (i.e., you ran `make services` from the top level folder) and now you want to see how it all fits together.
```
Comment (jschaul) [inline deploy/services-demo/README.md]: Perhaps you could add a line or two explaining that this is a "demo" setup (all data stored in-memory only)?

File: deploy/services-demo/demo.sh [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+USAGE="$0 [test-executable args...]"
+EXE=$1
+TOP_LEVEL="$( cd "$( dirname "${BASH_SOURCE[0]}" )/../.." && pwd )"
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+DIR="${TOP_LEVEL}/services"
+PARENT_PID=$$
+rm -f /tmp/demo.* # remove previous temp files, if any
+EXIT_STATUS_LOCATION=$(mktemp "/tmp/demo.XXXXXXXXXXX")
+echo 1 >${EXIT_STATUS_LOCATION}
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply deploy/services-demo/demo.sh]: 👍

File: deploy/services-demo/demo.sh [hunk 5]
```diff
@@ -0,0 +1,113 @@
+# You can use this script to simply start all services ./demo.sh
+
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply deploy/services-demo/demo.sh]: 👍

File: deploy/services-demo/demo.sh [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        || { echo "Not all services are compiled. How about you run 'cd ${TOP_LEVEL} && make services' first?"; exit 1; }
+}
+
+blue=6
+white=7
+green=10
+orange=3
+yellow=11
+purpleish=13
+dark_blue=17
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply deploy/services-demo/demo.sh]: 👍

File: services/nginz/Makefile [hunk 7]
```diff
@@ -69,17 +78,27 @@ $(DIST):
...
-	gpg --verify $(TMPDIR)/$(BUNDLE).asc $(TMPDIR)/$(BUNDLE)
-	tar zxf $(TMPDIR)/$(BUNDLE)
-	mv nginx-$(NGINX_VERSION) src
+conf/zwagger-ui/swagger-ui: $(TMPDIR)/$(SWAGGER_BUNDLE)
+	tar zxf $(TMPDIR)/$(SWAGGER_BUNDLE)
+	mv -vn swagger-ui-$(SWAGGER_VERSION)/dist conf/zwagger-ui/swagger-ui
+	rm -rf swagger-ui-$(SWAGGER_VERSION)
+
+$(TMPDIR)/$(SWAGGER_BUNDLE):
+	(cd $(TMPDIR); curl -L https://github.com/swagger-api/swagger-ui/archive/v$(SWAGGER_VERSION).tar.gz -o $(SWAGGER_BUNDLE))
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/nginz/Makefile]: Hmmm, this file is there once `swagger` is unpacked; not getting that on my setup, let's have a look at it later

File: deploy/services-demo/README.md (hunk 3 above)
Comment (tiago-loureiro) [reply deploy/services-demo/README.md]: Good idea, yes.

**Comment by fisx** [comment]
1. I started a branch `full-config-fisx` on top of this with changes (seemed easier for everybody than commenting here, feel free to ignore / cherry-pick).

2. During `make services`, I am asked to trust this pgp key, or rather get an error because I don't:

```
(1)	Maxim Dounin <mdounin@mdounin.ru>
	  2048 bit RSA key 520A9993A1C052F8, created: 2011-11-27
```

Is this mentioned in some other REAMDE?  Should it be?

```
$ git grep 520A9993A1C052F8
services/nginz/Dockerfile:    export GPG_KEYS=B0F4253373F8F6F510D42178520A9993A1C052F8 \
```

**Comment by fisx** [comment]
```
make -C src
make[2]: Entering directory '/home/mf/src/wire-server/services/nginz/src'
make -f objs/Makefile
make[3]: Entering directory '/home/mf/src/wire-server/services/nginz/src'
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -std=gnu99  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs \
        -o objs/src/core/nginx.o \
        src/core/nginx.c
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -std=gnu99  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs \
[...]
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -std=gnu99   -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/nginx-zauth-module/zauth_module.o \
        ../third_party/nginx-zauth-module/zauth_module.c
../third_party/nginx-zauth-module/zauth_module.c:6:19: fatal error: zauth.h: No such file or directory
 #include <zauth.h>
```

(If this is not related to this PR, let me know and I'll open an issue.)

**Comment by jschaul** [comment]
@fisx The gpg key import, and the need to install libzauth is mentioned [here](https://github.com/wireapp/wire-server/pull/283/files#diff-77e3ef28afcc92e23cd11ed21fe6b7bf), but perhaps the nginz documentation should be more explicit; or instead of saying "run `make services`", the top-level README should say "read services/nginz/README.md, install necessary dependencies, then run `cd services/nginz && make`"?

**Pull request merged**